tests/general/MementoBearTest.py

```diff
@@ -0,0 +1,289 @@
+
+    resp = requests.Response()
+    resp.raw = io.BytesIO()
+    resp.request = request
+    resp.status_code = status_code
+    return resp
+
+
+def memento_archive_status_mock(m, url, is_archived=True, override_head=True):
+    """
+    Mock the memento result for the given url.
+
+    :param m:             A `request_mock.Mocker()` instance.
+    :param url:           The ``url`` to be mocked.
+    :param is_archived:   True means the url will be marked as
+                          archived.
+    :param override_head: True means the url's head request will be overriden
+                          by this method.
+    """
+    if override_head:
+        m.head(url, headers={})
+    if is_archived:
+        # These two mocked responses will be received by memento_client
+        # and make memento_client returns a fake archived link.
+        # The first response headers will tell memento_client to redirect to
+        # the second one.
+        m.head(
+            'http://timetravel.mementoweb.org/timegate/%s' % url,
+            headers=memento_headers[0],
+            status_code=302)
+        m.head(
+            'http://web.archive.org/web/20170421070937/'
+            'https://www.google.com/',
+            headers=memento_headers[1],
+            status_code=200)
+    else:
+        m.head(
+            'http://timetravel.mementoweb.org/timegate/%s' % url,
+            status_code=404)
+
+
+def generate_redirects(m, url, redirect_length=1):
+    """
+    Generate redirects path for the given url.
+
+    :param m:               A `request_mock.Mocker()` instance.
+    :param url:             The ``url`` to be redirected.
+    :param redirect_length: How many redirects will be generated.
+    """
+    m.head(
+        url,
+        headers={'Location': '%s/1' % url},
+        status_code=302)
+
+    if redirect_length > 1:
+        for i in range(redirect_length - 1):
+            m.head(
+                '%s/%s' % (url, str(i + 1)),
+                headers={'Location': '%s/%s' % (url, str(i + 2))},
+                status_code=302
+            )
+
+        m.head('%s/%s' % (url, str(redirect_length)),
+               status_code=200)
+    else:
+        m.head('%s/1' % url, status_code=200)
+
+
+class MementoBearTestPrerequisites(unittest.TestCase):
+
+    def test_check_prerequisites(self):
+        with requests_mock.Mocker() as m:
+            m.add_matcher(custom_matcher)
+            self.assertTrue(MementoBear.check_prerequisites())
+
+
+class MementoBearTest(LocalBearTestHelper):
+    """
+    The tests are mocked (don't actually connect to internet) and
+    return the int conversion of the last three chars of
+    the URL as status code.
+
+    Check ``custom matcher`` for more info on implementation.
+    """
+
+    def setUp(self):
+        self.section = Section('')
+        MementoBear.check_prerequisites = lambda *args: True
+        self.uut = MementoBear(self.section, Queue())
+
+    def test_run(self):
```

Comment:
split this into several test methods with descriptive method names, each testing different things